requirements/AUR-CBI-INT-0070/standalone-project/src/application/main.c

```diff
@@ -0,0 +1,266 @@
+                controller_status->battery_capacity, controller_status->solar_panel_voltage,
+                controller_status->solar_panel_current);
+    }
+}
+
+void vm_controller_get_battery_capacity(asn1SccPID pid, void *capacity, size_t *capacity_size)
+{
+    asn1SccAuroraReal *capacity_value = (asn1SccAuroraReal *)capacity;
+    *capacity_value = battery_capacity;
+    *capacity_size = sizeof(asn1SccAuroraReal);
+}
+
+void vm_controller_get_payload_status(
+        asn1SccPID pid, void *temperature, size_t *temperature_size, void *flow_rate, size_t *flow_rate_size)
+{
+    asn1SccAuroraReal *temperature_value = (asn1SccAuroraReal *)temperature;
+    *temperature_value = payload_temperature;
+    *temperature_size = sizeof(asn1SccAuroraReal);
+    asn1SccAuroraReal *flow_rate_value = (asn1SccAuroraReal *)flow_rate;
+    *flow_rate_value = payload_flow_rate;
+    *flow_rate_size = sizeof(asn1SccAuroraReal);
+}
+
+void vm_controller_get_solar_panel_status(
+        asn1SccPID pid, void *voltage, size_t *voltage_size, void *current, size_t *current_size)
+{
+    asn1SccAuroraReal *voltage_value = (asn1SccAuroraReal *)voltage;
+    *voltage_value = solar_panel_voltage;
+    *voltage_size = sizeof(asn1SccAuroraReal);
+    asn1SccAuroraReal *current_value = (asn1SccAuroraReal *)current;
+    *current_value = solar_panel_current;
+    *current_size = sizeof(asn1SccAuroraReal);
+}
+
+void vm_controller_set_flow_controller(asn1SccPID pid, void *pwm, size_t pwm_size)
+{
+    if (pwm_size != sizeof(asn1SccAuroraReal)) {
+        printf("Invalid data in set_flow_controller\n");
+    } else {
+        payload_target_flow_rate = *((asn1SccAuroraReal *)pwm);
+        flow_rate_value = payload_target_flow_rate;
+    }
+}
+
+void vm_controller_set_power_point(asn1SccPID pid, void *pwm, size_t pwm_size)
+{
+    if (pwm_size != sizeof(asn1SccAuroraReal)) {
+        printf("Invalid data in set_power_point\n");
+    } else {
+        solar_panel_pwm = *((asn1SccAuroraReal *)pwm);
+    }
+}
+
+asn1SccPID controller_step_get_sender()
+{
+    return PID_master_clock;
+}
+
+asn1SccPID algorithms_calculate_mppt_get_sender()
+{
+    return PID_controller;
+}
+
+asn1SccPID algorithms_calculate_pid_get_sender()
+{
+    return PID_controller;
+}
+
+// application
+void init_application()
+{
+    srand(time(NULL));
+    flow_rate_value = 0.0;
+    battery_capacity = 0.48;
+    solar_panel_voltage = 9.3;
+    solar_panel_current = 1.3;
+    solar_panel_pwm = 0.0;
+
+    electric_power_supply_step_count = 0;
+    electric_power_supply_amp_40 = 0.0;
+    electric_power_supply_amp_20 = 0.0;
+    electric_power_supply_amp_10 = 0.0;
+    electric_power_supply_amp_5 = 0.0;
+
+    payload_temperature = 18.0;
+    payload_flow_rate = 4.0;
+    payload_target_flow_rate = 0.0;
+
+    payload_step_count = 0;
+    payload_amp_100 = 0.0;
+    payload_amp_50 = 0.0;
+    payload_amp_25 = 0.0;
+}
+
+void application_step()
+{
+    if (electric_power_supply_step_count == 0) {
```

Comment:
It would be really nice if this function was refactored into 3 or 4 smaller ones, dealing with EPS and Payload separately.

Reply:
Done